Upgrade ROR API From V1 to V2

momo3404 opened the pull request:
Fixes [#3588](https://github.com/DMPRoadmap/roadmap/issues/3588).

Changes proposed in this PR:
- Update `ror_service.rb` and `ror_service_spec.rb` to work with V2 of the ROR API as V1 has been officially deprecated.
- The full list of changes from V1 to V2 can be found [here](https://ror.readme.io/docs/schema-v2#/). The main changes that affect DMP Assistant are the changes in how names, languages, websites, locations, acronyms, and external IDs are handled in the schema.

momo3404 committed:
Update ROR configuration to use V2 API

- The base url "https://api.ror.org/" defaults to the V2 API

momo3404 committed:
Update `parse_results` in `ror_service` and add helper functions

- In the V2 API, names now returns an array of names. The display_name helper function fetches the display name.
- In the V2 API, there are now multiple links which can point to more than just the org's website. The helper function org_url retrieves the org's website.
- parse_results is edited to use these helper functions in order to parse results properly.
- Also fixed a comment typo.

momo3404 committed:
Update `org_name` in `ror_service`

- The function is updated to use the display_name helper function and comply with the V2 API schema.

momo3404 committed:
Update `org_language` in `ror_service`

- Languages were simplified in the v2 api and can now be extracted from names.
- org_language was updated to do that.

momo3404 committed:
Update `org_website` in `ror_service`

- org_website was updated to use the org_url helper function.

momo3404 committed:
Update `fundref_id` in `ror_service`

- The external_ids schema was changed in the V2 API to use "type". fundref_id was updated to reflect this change, and simplify the logic of retrieving preferred and all ids.

momo3404 committed:
Update API Response tests in `ror_service_spec`

- Update tests to follow new schema regarding names, external ids, and links.

momo3404 committed:
Update `query_ror`, `process_pages`, `parse_results` tests in `ror_service_spec`

momo3404 committed:
Update `org_name`, `org_website`, `fundref_id` tests in `ror_service_spec`

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -135,20 +135,47 @@ def process_pages(term:, json:, filters: [])
+      #     {"lang": "en", "types": ["ror_display","label"], "value": "Harvard University"},
+      #     {"lang": "es","types": ["label"], "value": "Universidad de Harvard"}
+      # ]
+      def display_name(item)
```
Just a little tidying thing; it might make things a little more organised to just move  `def display_name(item)` and `def org_url(item)` below `def parse_results(json:)` to group them with the other parse_results helper methods.

aaronskiba replied:
Renaming `def display_name` to `def sort_name` might be worthwhile as well.

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -135,20 +135,47 @@ def process_pages(term:, json:, filters: [])
-            sort_name: item['name'],
-            url: item.fetch('links', []).first,
+            sort_name: name,
+            url: org_url(item)&.dig('value'),
```
`url: org_url(item)&.dig('value'),` is here and `link = org_url(item)&.fetch('value', nil)` is in `def org_website`. It might be cleaner to just have `def org_url` return the desired value.

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -163,48 +190,71 @@ def parse_results(json:)
+      #   ]
+      # }
+      def org_language(item:) # rubocop:disable Metrics/CyclomaticComplexity
         dflt = I18n.default_locale || 'en'
```
This isn't related to your changes, but `dflt = I18n.default_locale || 'en'` worries me a bit.
```ruby
irb(main):001> I18n.default_locale
=> :"en-CA"
irb(main):002> I18n.default_locale.class
=> Symbol
```
https://ror.readme.io/docs/ror-data-structure states "Each name variant has a corresponding language tag with the two-letter [ISO-639](https://www.loc.gov/standards/iso639-2/php/code_list.php) language code where available."

However, it might only be used in  `app/services/org_selection/hash_to_org_service.rb`:

```ruby
      # Initialize a new Org from the hash
      def initialize_org(hash:)
        return nil unless hash.present? && hash[:name].present?

        Org.new(
          name: hash[:name],
          links: links_from_hash(name: hash[:name], website: hash[:url]),
          language: language_from_hash(hash: hash),
          target_url: hash[:url],
          institution: true,
          is_other: false,
          abbreviation: abbreviation_from_hash(hash: hash)
        )
      end

      # Get the language from the hash or use the default
      def language_from_hash(hash:)
        return Language.default unless hash.present? && hash[:language].present?

        Language.where(abbreviation: hash[:language]).first || Language.default
      end
```
If that's the case, then there may not actually be an issue.

momo3404 replied:
I might be wrong, but if `Language.default` resolves to a two letter string, then it might indeed not be an issue. Now that I think more about that code, would it be better to change the first two lines of `org_language` to `return 'en' unless item.present?`

aaronskiba replied:
Yes, 'en' works for us, but it might be even better to make it more generalizable if we end up contributing this to DMPRoadmap.

I have an idea for this. Maybe I'll make the commit in a branch copied from here.

momo3404 replied:
You can just let me know the idea, I have time to implement it.

aaronskiba replied:
You could probably just make the following change in `def org_language(item:)`:
```ruby
        # ROR uses two-letter / ISO 639-1 language codes (e.g. "en")
        dflt = Language.default.abbreviation.split('-').first
```

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -135,20 +135,47 @@ def process_pages(term:, json:, filters: [])
+      # ]
+      def org_url(item)
+        # "links" may be nil, a non‑array, or the whole item may not be a Hash
+        return nil unless item.is_a?(Hash)
```
I think we can remove `return nil unless item.is_a?(Hash)`. We are already using it as a hash before this code executed.

aaronskiba replied:
Also, the comment states `# "links" may be nil, a non‑array, or the whole item may not be a Hash`, but https://ror.readme.io/docs/ror-data-structure states that "links must be an array (possibly empty).

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -163,48 +190,71 @@ def parse_results(json:)
+        external_ids = item['external_ids']
+        return '' unless external_ids.present?
+
+        fundref = external_ids.find { |id| id['type'].to_s.casecmp('fundref').zero? }
```
We can remove the `.to_s` and it looks like `.casecmp('fundref').zero?` can maybe be simplified to `.casecmp?('fundref')`. But after looking at https://ror.readme.io/docs/fields the following might be sufficient:
```ruby
fundref = external_ids.find { |id| id['type'] == 'fundref' }
```

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -163,48 +190,71 @@ def parse_results(json:)
-        return '' unless item.present? && item['external_ids'].present?
-        return '' unless item['external_ids'].fetch('FundRef', {}).any?
+        external_ids = item['external_ids']
+        return '' unless external_ids.present?
```
I wonder if some of this code is overly defensive. There are a number of checks like this throughout the code. However, in some cases, such as this one, the API contract implies that such a check isn't necessary (e.g. https://ror.readme.io/docs/ror-data-structure states that 'external_ids' will return an array (possibly empty).

momo3404 replied:
A lot of this defensive behaviour was not introduced in this PR. The external_ids check, for instance, was already present in the existing code, and removing it causes multiple failures in ror_service_spec. Given that, I think keeping the check is reasonable here, unless we want to also want to rewrite the tests, which might be more work than simply leaving this behaviour untouched.

aaronskiba replied:
Thanks for looking into that. For external_ids, even v1 appears to guarantee the presence of the field in the schema (https://ror.readme.io/docs/schema-v1), so I'm not sure why this guard was ever added in the first place.

Given the published contract, checks like `return '' unless external_ids.present?` are redundant unless we can find an actual violation in the ROR data. Rather than preserving the legacy defensiveness/redundancy, I’d lean toward taking the extra step to simplify the code and update the specs accordingly (Boy Scout Rule: https://deviq.com/principles/boy-scout-rule).

momo3404 replied:
Reworking the tests to satisfy this turned out to not actually be complicated, so I pushed that change and squashed it with the commit for simplifying the fundref case comparison. Should work nicely now, let me know if that's good.

aaronskiba reviewed (commented):
I wonder if some of this code may be overly defensive. There are several checks around key presence, types, and values that appear to be guaranteed by the ROR v2 data structure contract (https://ror.readme.io/docs/ror-data-structure).

If we trusted that ROR is honoring its published schema, it could allow us to simplify the code.

At the same time, if there are in fact any schema violations, we could just log them as follows in `parse_results`:

```ruby
rescue StandardError => e
  Rails.logger.error("ROR schema violation: #{e.message}", item:)
end
```

momo3404 committed:
Improve `org_url` and `display_name` in `ror_service`

- Remove the unnecessary hash check for items since links must be an array
- Change org_url return value from links to website url to simplify use
- Rename `display_name` to `sort_name`
- Move functions below parse_results

momo3404 committed:
Simplified fundref and external_ids handling

- Removed the fundref case comparison since contract ensures it will always be 'fundref'
- Removed the unnecessary all.first.present check
- Removed redundant external_ids presence check since external_ids is guaranteed to be always present as an array, even if empty
- Edited ror_service_spec tests to work with `fundref` and empty external_ids

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -158,53 +159,99 @@ def parse_results(json:)
+      # }
+      def org_language(item:) # rubocop:disable Metrics/CyclomaticComplexity
         dflt = I18n.default_locale || 'en'
         return dflt unless item.present?
```
`return dflt unless item.present?`

This looks like another redundant check in the legacy code (`def parse_results` has established that `item.present?` is true prior to this code). We could probably just remove this to clean up the code.

aaronskiba replied:
`external_ids = item['external_ids'] if item` and `links = item['links'] if item` might be redundant checks as well.

momo3404 replied:
For external_ids and links, could there not be a case where the functions are called with an actual 'nil' item?

aaronskiba replied:
Good question/concern. We have the methods set as private, so that limits the concern a lot. Technically, other RorService methods could call these, but there is no existing code for it and I don't foresee it being added.

Right now, `parse_results` is only ever called by `process_pages` and `process_pages` is only ever called by `search`.  I think we're safe to trust that the `parse_results` helper methods will only ever be called by `parse_results`.

momo3404 replied:
Sounds good, I'll just have to remove one test to remove those checks.

momo3404 committed:
Remove redundant nil checks and rework `org_language`

- Edit dflt in org_language to reflect two letter codes
- Remove redundant 'if item' checks in org_url and fundref_id since parse_results already processes this
- Remove check for nil test for org_name and org_language since parse_results checks this condition already
- Edit array check in links since links will always be an array so the array check is reworked

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -158,53 +159,99 @@ def parse_results(json:)
       def org_name(item:)
-        return '' unless item.present? && item['name'].present?
+        name = sort_name(item)
+        return '' unless name.present?
```
I think `return '' unless name.present?` can be removed here. The following in `parse_results` is already performing the equivalent check:
```ruby
        json['items'].each do |item|
          name = sort_name(item)
          next unless item['id'].present? && name.present?
```

aaronskiba committed:
Fix ROR <-> DB language mapping

app/services/org_selection/hash_to_org_service.rb
`language_from_hash` previously tried to match ROR two-letter codes (e.g., "en") directly against `Language.abbreviation` (e.g., "en-CA"), causing it to always fall back to `Language.default`. This change fixes the lookup by matching against the `Language.abbreviation` language subtag (the portion before any region) and ordering by `default_language: :desc` to prioritize the default language when multiple matches exist.
- Added check for exact match in language_from_hash

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -158,53 +159,98 @@ def parse_results(json:)
+      #     {"lang": "en", "types": ["ror_display","label"], "value": "Harvard University"},
+      #     {"lang": "es","types": ["label"], "value": "Universidad de Harvard"}
+      # ]
+      def sort_name(item)
```
This might be more defensive than necesary. https://ror.readme.io/docs/fields states the following for `names.types`:

The type(s) associated with the name contained in names.value. Each name must have at least 1 type, and exactly 1 name must have ror_display in its types. Each name can have multiple types, for example ror_display and label.

momo3404 replied:
Would this be better? 
```ruby
def sort_name(item)
  item['names']
    .find { |n| n['types'].include?('ror_display') }['value']
end
```

aaronskiba replied:
Would it be better to use the `ror_display_entry(item)` idea you suggested in our other conversation?

momo3404 replied:
Yes, could definitely do that.

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -158,53 +159,98 @@ def parse_results(json:)
+      #     { "value": "University of Montreal", "types": ["alias"], "lang": "en" }
+      #   ]
+      # }
       def org_language(item:)
```
I guess https://github.com/portagenetwork/roadmap/pull/1247/files#r2643715808 also applies to `def org_language(item:)`.

Looking at the duplicate code between def `sort_name(item)` and `def org_language(item:)` maybe you could DRY the two up as well (`def org_name(item:)` might benefit from this as well). Maybe something like the following would work?

```ruby
def parse_results(json:)
  ror_display = ror_display_entry(item) # This will replace name = sort_name(item)
  ...
  name: org_name(item: item, ror_display: ror_display),
  sort_name: sort_name(ror_display),
  ...
  language: org_language(ror_display),
```

```ruby
def ror_display_entry(item)
  item['names'].find { |n| n['types'].include?('ror_display') }
end
```
The other referenced helpers would need small updates as well.

momo3404 replied:
I'm thinking a simpler way of doing this would be to just use the ror_display helper in the helper functions instead of parse_results, so it would look something like this 

```ruby
def org_language(item:)
  # ROR uses two-letter / ISO 639-1 language codes (e.g. "en")
  dflt = I18n.default_locale.to_s.split('-').first

  ror_display_entry(item)&.fetch('lang', dflt) || dflt
end
``` 
I think this achieves the DRY goal without overcomplicating `parse_results`.

aaronskiba replied:
That sounds okay. With that approach, every helper would be re-executing `ror_display` helper. But that might just be a trivial for performance. I think your idea is cool and I like the idea of simplifying parse_results. I have one other suggestion if we go with what you're proposing:
```ruby
name = sort_name(item)
next unless item['id'].present? && name.present?
```
I worry about how the legacy code would just skip to the next item if the condition is false. `item['id'].present? && name.present?` should be guaranteed by the ROR API contract. So if this condition is in fact false, it may signify that something is wrong with our code. Rather than silently iterating to the next item, it could log an error:

```ruby
  json['items'].each do |item|
    begin
      results << {
        ror: item.fetch('id').gsub(/^#{landing_page_url}/, ''),
        name: org_name(item: item),
        sort_name: sort_name(item),
        language: org_language(item),
        url: org_url(item),
        fundref: fundref_id(item: item),
        abbreviation: item.fetch('acronyms', []).first
      }
    rescue KeyError, NoMethodError => e
      # Log error and continue to the next item
      Rails.logger.error("Invalid ROR record: #{e.message}, item: #{item.inspect}")
    end
```

momo3404 replied:
That edit to `parse_results` unfortunately doesn't catch the error, it just fills the values with nil. This is what the failing test indicated:

```ruby
it 'ignores items with no name or id' do
        json = { items: [
          { id: Faker::Internet.url, names: [{ types: ['ror_display'], value: Faker::Lorem.word }], external_ids: [] },
          { id: Faker::Internet.url, names: [], external_ids: [] },
          { names: [{ types: ['ror_display'], value: Faker::Lorem.word }], external_ids: [] }
        ] }.to_json

        items = described_class.send(:parse_results, json: JSON.parse(json))
        byebug
        expect(items.length).to eql(1)
      end
```

```
(byebug) items
[{:ror=>"http://block.test/mellisa", :name=>"suscipit", :sort_name=>"ExternalApis::RorService", :url=>nil, :language=>"en", :fundref=>"", :abbreviation=>nil}, {:ror=>"http://torphy.example/mignon", :name=>nil, :sort_name=>"ExternalApis::RorService", :url=>nil, :language=>"en", :fundref=>"", :abbreviation=>nil}]
```

I think keeping the check might be okay if the error logging is added as a safeguard.

aaronskiba replied:
Hmm, I'm still leaning toward removing that check. Those values are explicitly guaranteed by the ROR API spec. Also, `next unless item['id'].present? && name.present?` doesn't just ignore "bad data", it actively suppresses more useful errors. For instance, this example shows how it could suppress a KeyError:

```
irb(main):001> item = {}
=> {}
irb(main):002> item['id'].present?
=> false
irb(main):003> item.fetch('id').gsub('test', '')
(irb):3:in `fetch': key not found: "id" (KeyError)
        from (irb):3:in `<main>'
irb(main):004>
```

momo3404 replied:
Let me know what you think of this.

```ruby
json['items'].each do |item|
      id = item&.fetch('id')
      name = sort_name(item)
      raise KeyError, 'Missing ror_display name' if name.blank?

      results << {
        ror: id.gsub(/^#{landing_page_url}/, ''),
        name: org_name(item: item),
        sort_name: name,
        url: org_url(item),
        language: org_language(item: item),
        fundref: fundref_id(item: item),
        abbreviation: item.fetch('acronyms', []).first
      }
    rescue KeyError, NoMethodError => e
      Rails.logger.error("Invalid ROR record: #{e.message}, item: #{item.inspect}")
    end
 ```
It includes error logging to prevent silent errors while also preventing malinformed items from being passed into `results`.

aaronskiba replied:
I don't really see an advantage to that. I'd maybe just remove the safe operator from line 172 (https://github.com/portagenetwork/roadmap/pull/1247#discussion_r2645965131) and catch the error in the bottom rescue block.

momo3404 replied:
Sounds good, I think this should work then. 

```ruby
def parse_results(json:)
        results = []
        return results unless json.present? && json.fetch('items', []).any?

        json['items'].each do |item|
          name = sort_name(item: item)

          results << {
            ror: item['id'].gsub(/^#{landing_page_url}/, ''),
            name: org_name(item: item),
            sort_name: name,
            url: org_url(item: item),
            language: org_language(item: item),
            fundref: fundref_id(item: item),
            abbreviation: item.fetch('acronyms', []).first
          }
        rescue KeyError, NoMethodError => e
          Rails.logger.error(
            "Invalid ROR record: #{e.class} - #{e.message}, item: #{item.inspect}"
          )
        end
        results
      end
```

momo3404 committed:
Add `ror_display_entry` helper function

- This change helps clean up repeated code and simplifies sort_name handling
- The test variable was removed since the API contract ensures 'ror_display' will always be there

ror_entry helper

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -158,53 +159,99 @@ def parse_results(json:)
+      #     {"lang": "es","types": ["label"], "value": "Universidad de Harvard"}
+      # ]
+      def sort_name(item)
+        ror_display_entry(item)&.fetch('value', nil)
```
We could just use `ror_display_entry(item)['value']` here.
Alternatively, `ror_display_entry(item).fetch('value')` should fail loud and catch the error you said was being missed here https://github.com/portagenetwork/roadmap/pull/1247/files#r2644246570

momo3404 replied:
I'll go with `.fetch('value')`, a test revealed `['value']` would throw a nil error when there's no id or name so that's why I initially didn't go with it.

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -142,13 +142,14 @@ def parse_results(json:)
+            url: org_url(item),
             language: org_language(item: item),
             fundref: fundref_id(item: item),
             abbreviation: item.fetch('acronyms', []).first
```
This is small, but all of the existing methods use kwargs, (`(item: item)`), whereas all of our new methods are using positional args (`(item)`). It might be better to just update the new methods to also use kwargs.

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -158,53 +159,99 @@ def parse_results(json:)
 
-        labels = item.fetch('labels', [{ iso639: dflt }])
-        labels.first&.fetch('iso639', I18n.default_locale) || dflt
+        ror_display_entry(item)&.fetch('lang', dflt) || dflt
```
If `ror_display_entry(item)` is guaranteed by the API contract, then I'd remove the safe navigation operator here.

momo3404 replied:
I'll have to remove another test to do this, just double checking since we've already removed a good amount with the idea of trusting the API contract.

aaronskiba replied:
I think it's okay to remove the test, if that is necessary for these changes.

aaronskiba replied:
It doesn't have to be perfect, but it would be nice to clean the branch up as we've made these tweaks. I imagine some commits could at least be squashed together in some spots.

aaronskiba commented on app/services/org_selection/hash_to_org_service.rb:
```diff
@@ -125,7 +125,15 @@ def abbreviation_from_hash(hash:)
 
-        Language.where(abbreviation: hash[:language]).first || Language.default
+        # Try exact match first
+        exact = Language.find_by(abbreviation: hash[:language])
```
This is a good addition. Roadmap has some locales with the two-letter language codes and this is good for priortising them.

momo3404 committed:
Update helper functions to use keyword arguments

- This matches the behaviour of other functions in the file

momo3404 committed:
Update parse_results to log errors instead of skipping

- remove item id and name checks since API contract ensures they're present
- add error logging to make sure errors are no longer supressed
- edit test to make sure error is being logged

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -142,69 +142,119 @@ def parse_results(json:)
             name: org_name(item: item),
-            sort_name: item['name'],
-            url: item.fetch('links', []).first,
+            sort_name: name,
```
We could probably just go `sort_name: sort_name(item: item)` now and remove the assignment.

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -142,69 +142,117 @@ def parse_results(json:)
             fundref: fundref_id(item: item),
             abbreviation: item.fetch('acronyms', []).first
           }
+        rescue KeyError, NoMethodError => e
+          Rails.logger.error(
+            "Invalid ROR record: #{e.class} - #{e.message}, item: #{item.inspect}"
+          )
```
It might be good to just add a test that verifies the rescue behaviour.

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -142,69 +142,117 @@ def parse_results(json:)
 
-        labels = item.fetch('labels', [{ iso639: dflt }])
-        labels.first&.fetch('iso639', I18n.default_locale) || dflt
+        ror_display_entry(item: item).fetch('lang', dflt) || dflt
```
https://ror.readme.io/docs/ror-data-structure#names

"Each name variant has a corresponding language tag with the two-letter [ISO-639](https://www.loc.gov/standards/iso639-2/php/code_list.php) language code where available. The language is not known for all name variants, so corresponding language codes may be null for some names. ROR name metadata fields support multiple languages and character sets."

Based on the above, it sounds like `lang` may have a null value, but we can trust that the key exists. We could probably remove the missing key fallback here: (i.e. change it to `ror_display_entry(item: item).fetch('lang') || dflt`).

momo3404 replied:
I think I'd prefer to keep the fallback. Relying on this behavior assumes the schema will never omit `lang` in future updates or edge cases. Keeping the fallback adds a small layer of resilience at little cost, especially when handling data that might evolve.

aaronskiba replied:
Cool, we can keep it.

What are your thoughts on `.fetch('lang', dflt) || dflt`? I guess it comes from the legacy code, but it feels a bit awkward to read. I think the following refactor could improve that:
`ror_display_entry(item: item)['lang']|| dflt`
It seems like it will behave as desired:
```
irb(main):001> item = {}
=> {}
irb(main):002> item['lang'] || I18n.default_locale
=> :"en-CA"
irb(main):003> item['lang'] = nil
=> nil
irb(main):004> item
=> {"lang"=>nil}
irb(main):005> item['lang'] || I18n.default_locale
=> :"en-CA"
```

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -142,69 +142,117 @@ def parse_results(json:)
+      def org_url(item:)
+        links = item['links']
+        # links must be an array
+        return nil unless links.is_a?(Array)
```
https://ror.readme.io/docs/ror-data-structure#fields
The array may be empty, but I think we can trust that `item['links']` itself is an array. `return nil unless links.is_a?(Array)` might be redundant.

momo3404 replied:
I think reworking this to `return nil unless links.present?` should work

aaronskiba replied:
I think just removing the array check should also work:
```
$ rails c
Running via Spring preloader in process 1596074
Loading development environment (Rails 6.1.7.9)
irb(main):001> links = []
=> []
irb(main):002> links.find { |l| l['type'] == 'website' }&.fetch('value', nil)
=> nil
```

aaronskiba replied:
While we're at it, I think we could maybe even update the bottom line to the following:
`links.find { |l| l['type'] == 'website' }&.fetch('value')`
Although, `links == []` is possible, if it is not empty, I think each of the containing objects should have the `value` key.

momo3404 replied:
I made that change and squashed the commit into the 'remove redundant nil checks' commit

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -142,69 +142,117 @@ def parse_results(json:)
-        return '' unless item.present? && item['name'].present?
+        name = sort_name(item: item)
 
         country = item.fetch('country', {}).fetch('country_name', '')
```
`country = item.fetch('country', {}).fetch('country_name', '')`

This looks like it might need updating. Schema v1.0 -> v2.0 removed `country` as a top-level field.

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -142,69 +142,117 @@ def parse_results(json:)
+            url: org_url(item: item),
             language: org_language(item: item),
             fundref: fundref_id(item: item),
             abbreviation: item.fetch('acronyms', []).first
```
`abbreviation: item.fetch('acronyms', []).first`

This looks like it will need updating. https://ror.readme.io/docs/schema-v2 shows how `acronyms` is no longer a top-level field.

aaronskiba commented on spec/services/external_apis/ror_service_spec.rb:
```diff
@@ -71,28 +71,32 @@
           items: [
```
This relates to comment https://github.com/portagenetwork/roadmap/pull/1247/files#r2669703045.

I'd double check this structure against https://ror.readme.io/docs/ror-data-structure#example-record:

As one example, we have the following:
`country: { country_name: 'United States', country_code: 'US' },`

But the readme has the following:
```
   "locations" : [
      {
         "geonames_details" : {
            "continent_code" : "EU",
            "continent_name" : "Europe",
            "country_code" : "DE",
            "country_name" : "Germany",
            "country_subdivision_code" : "NW",
            "country_subdivision_name" : "North Rhine-Westphalia",
            "lat" : 51.45657,
            "lng" : 7.01228,
            "name" : "Essen"
         },
         "geonames_id" : 2928810
      }
   ],
```

aaronskiba replied:
It looks like acronyms will have to be fixed here too.

momo3404 replied:
Just updated the test schemas that use acronyms, the helper function works too

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -142,69 +142,117 @@ def parse_results(json:)
         return results unless json.present? && json.fetch('items', []).any?
```
Maybe just a personal preference, but your call. It would be cleaner to remove the JSON structure comments and just replace them with a link to https://ror.readme.io/docs/ror-data-structure above `def parse results`.

momo3404 replied:
I see the point in doing that, but I do think having the comments there make it easier for someone to immediately understand what the code is doing. It could also be helpful in the future if there's another update to the schema, as it would allow direct comparison between the two schemas for the dev working with the code.

momo3404 committed:
Update locations in `ror_service` and `ror_service_spec`

- The V2 ROR schema nests country names within locations
- The helper function org_country was added to extract country name from locations
- Tests were also updated to match this new schema

momo3404 committed:
Update abbreviation in `ror_service` and `ror_service_spec`

- The V2 schema moves the acronym value inside the names array. - The org_abbreviation helper is added to extract that value for parse_results
- Update acronyms in tests

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -142,69 +142,132 @@ def parse_results(json:)
+
+      # Returns the country name from locations
+      # "locations" : [ { "geonames_details" : { "country_name" : "Germany",}, "geonames_id" : 2928810 } ]
+      def org_country(item:)
+        location = item.fetch('locations', []).find { |l| l['geonames_details'].present? }
+        location&.dig('geonames_details', 'country_name') || ''
+      end
+
+      # Extract acronym/abbreviation from names
+      # "names" : [ { "value" : "UC", "types": ["acronym"], "lang" : "en" } ]
+      def org_abbreviation(item:)
+        item.fetch('names', [])
+            .find { |n| n['types']&.include?('acronym') }
+            &.fetch('value')
+      end
```
This is looking better. If we are trusting ROR's API contract, we might be able to simplify these a little bit. We should maybe only use `&.` when something may legitimately be absent.

`locations` and `names` are both required top-level fields that contain arrays, so we probably remove the failbacks on each of those `.fetch()` calls (i.e. just use `item.fetch('locations')` and `item.fetch('names')`).

You could probably remove the safe navigation operator in `n['types']&.include?('acronym'`) as well (doing so would also make it more consistent with assumptions in `def ror_display_entry(item:)`).

Also, `nil.to_s == ''`. Up to you, but you could even use `location&.dig('geonames_details', 'country_name').to_s` along with any other changes here.

momo3404 replied:
I made those changes and updated the tests accordingly. Squashed it into the abbreviation and location commits

aaronskiba replied:
Cool. I'm sorry, but maybe something went wrong with the squash?

I'm not seeing those changes, and it looks like `&.fetch('value')` was changed to `.fetch('value')` for 'acronym'. This is one where we may in fact really need the safe navigation operator.

momo3404 replied:
Sorry, must've misunderstood the comment. Is this good? 

```ruby
def org_country(item:)
        location = item.fetch('locations', []).find { |l| l['geonames_details'].present? }
        location.dig('geonames_details', 'country_name') || ''
      end

      # Extract acronym/abbreviation from names
      # "names" : [ { "value" : "UC", "types": ["acronym"], "lang" : "en" } ]
      def org_abbreviation(item:)
        item.fetch('names', [])
            .find { |n| n['types'].include?('acronym') }
            &.fetch('value')
      end
```

aaronskiba replied:
That is good for fixing the `&.fetch`. I might even do the following to keep it consistent with what we've already written in `def ror_display_entry(item:)`:
```ruby
def org_abbreviation(item:)
  item['names'].find { |n| n['types'].include?('acronym') }&.fetch('value')
end
```

`def org_country(item:)` is a little worrying. If `item.fetch('locations', [])` returns `[]` then `location.dig('geonames_details', 'country_name') || ''` will raise. If you want to keep it as safe as possible, I'd recommend something like this:
```ruby
def org_country(item:)
        location = item['locations']&.find { |l| l['geonames_details'].present? }
        location&.dig('geonames_details', 'country_name').to_s
      end
```

momo3404 replied:
Just made both those changes and squashed into the location and abbreviation commits.

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -142,69 +142,132 @@ def parse_results(json:)
+        links = item['links']
+        # links must not be empty
+        return nil unless links.present?
+
+        links.find { |l| l['type'] == 'website' }&.fetch('value')
+      end
```
Maybe we could make a small cleanup here. If links is an empty array, `links.find { |l| l['type'] == 'website' }&.fetch('value')` will return nil, so we don't need separate check on line 184. You could even just make this a one-liner (e.g. `item['links'].find { |l| l['type'] == 'website' }&.fetch('value')`)

momo3404 replied:
`[ ].find { … }` does return nil as expected, but I do once again think putting too much trust in this part of the schema not changing might not be safe, since it could easily be changed to where links may be `nil` or fit into a different part of the schema, in which case a `NoMethodError` will be raised if we don't have the early return.

We could still make this a one-liner though with `item['links']&.find { |l| l['type'] == 'website' }&.fetch('value')`.

aaronskiba replied:
I guess the other option is letting it fail loudly and handling it via the rescue block. I don't think we'd have to worry about such schema changes until another major API version change (i.e. v3).

At the same time, it's just a little extra `&.`, which is no big deal. And that would at least protect us from potentially malformed data. I'll leave it up to you. 👍

momo3404 replied:
I'll go with `item['links']&.find { |l| l['type'] == 'website' }&.fetch('value')`, seems like a good fit for cleanliness and security.

momo3404 committed:
Update org_url to use one line operator

aaronskiba committed:
Rework `language_from_hash` and `org_language`

- Edit language_from_hash to ensure proper query sanitization and to merge the two former queries into one
- Edit org_language to be cleaner with a one line operation

momo3404 committed:
Add changelog entry

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -142,69 +142,123 @@ def parse_results(json:)
-        return nil if item['links'].first.blank?
+        link = org_url(item: item)
+        return nil unless link.present?
 
         # A website was found, so extract just the domain without the www
         domain_regex = %r{^(?:http://|www\.|https://)([^/]+)}
-        website = item['links'].first.scan(domain_regex).last.first
+        website = link.scan(domain_regex).last.first
         website.gsub('www.', '')
```
Your refactor is correct, but there is a potential issue with the legacy code itself:
https://ror.readme.io/docs/fields states the following about `links.value`:
"Valid URI, according to [IETF RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986)"

IETF RFC 3986 lists the following examples as valid URIs:
```
 ftp://ftp.is.co.za/rfc/rfc1808.txt
http://www.ietf.org/rfc/rfc2396.txt
ldap://[2001:db8::7]/c=GB?objectClass?one
mailto:John.Doe@example.com
news:comp.infosystems.www.servers.unix
tel:+1-816-555-1212
telnet://192.0.2.16:80/
urn:oasis:names:specification:docbook:dtd:xml:4.1.2
```

But the current code will throw a NoMethodError for some of these examples:
```
irb(main):001> link = ' ftp://ftp.is.co.za/rfc/rfc1808.txt'
=> " ftp://ftp.is.co.za/rfc/rfc1808.txt"
irb(main):002> domain_regex = %r{^(?:http://|www\.|https://)([^/]+)}
=> /^(?:http:\/\/|www\.|https:\/\/)([^\/]+)/
irb(main):003> website = link.scan(domain_regex).last.first
(irb):3:in `<main>': undefined method `first' for nil:NilClass (NoMethodError)
website = link.scan(domain_regex).last.first
                                      ^^^^^^
irb(main):004> website = link.scan(domain_regex)
=> []
```

I'd recommend the following as a minimal change:
```ruby
domain_regex = %r{^(? :http://|www\.|https://)([^/]+)}
matches = link.scan(domain_regex)
return nil if matches.empty?
website = matches.last.first
website.gsub('www.', '')
```

momo3404 replied:
Nice catch! Just added the commit.

aaronskiba commented on app/services/external_apis/ror_service.rb:
```diff
@@ -142,69 +142,123 @@ def parse_results(json:)
+      # ]
       def org_website(item:)
-        return nil unless item.present? && item.fetch('links', [])&.any?
-        return nil if item['links'].first.blank?
+        link = org_url(item: item)
+        return nil unless link.present?
 
         # A website was found, so extract just the domain without the www
         domain_regex = %r{^(?:http://|www\.|https://)([^/]+)}
-        website = item['links'].first.scan(domain_regex).last.first
+        website = link.scan(domain_regex).last.first
         website.gsub('www.', '')
       end
 
-      # Extracts the FundRef Id if available
+      # Extracts the FundRef Id from external ids if available
+      # "external_ids": [
+      #   {"type": "fundref", "preferred": "12345", "all": ["12345", "67890"]},
+      #   {"type": "SomeOtherID", "preferred": "501100000000", "all": ["501100000000"]}
+      # ]
       def fundref_id(item:)
-        return '' unless item.present? && item['external_ids'].present?
-        return '' unless item['external_ids'].fetch('FundRef', {}).any?
+        external_ids = item['external_ids']
+
+        fundref = external_ids.find { |id| id['type'] == 'fundref' }
+        return '' unless fundref.present?
 
         # If a preferred Id was specified then use it
-        ret = item['external_ids'].fetch('FundRef', {}).fetch('preferred', '')
-        return ret if ret.present?
+        preferred = fundref['preferred']
+        return preferred if preferred.present?
 
         # Otherwise take the first one listed
-        item['external_ids'].fetch('FundRef', {}).fetch('all', []).first
+        all = fundref['all']
+        return all.first if all.present?
+
+        ''
```
I'm going overboard with the refactors, but if you wanted you could replace the last few lines with `fundref['all']&.first.to_s`

momo3404 replied:
Lol no worries, I do think I'll keep the current code though. Even though it's a few more lines, it's a bit cleaner to read.

momo3404 committed:
Edit domain extraction logic in `org_website`

- The former logic threw a NoMethodError for some valid URIs, this fix ensures that nil is returned in case a match is not found

aaronskiba reviewed (approved):
Good work. We'll get a chance to further see the effects of these changes as we work on the contributors and funders PRs.
